edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/MqttBrokerModule.cs

```diff
@@ -0,0 +1,92 @@
+        }
+
+        protected override void Load(ContainerBuilder builder)
+        {
+            var componentTypes = MqttBridgeComponentDiscovery.GetCandidateTypes().ToArray();
+
+            // The classes will be registered by two types:
+            // 1) by its own type - that is because the connector is going to
+            //    request an instance by the type name as it finds ISubscriber, etc
+            //    derived types.
+            // 2) by every interface name that is not the following:
+            //      IMessageConsumer/IProducer/ISubscriber/ISubscriptionWatcher.
+            //    That is because some of the producers/consumers are also special services
+            //    and components will need those injected by custom interfaces.
+            builder.RegisterTypes(componentTypes)
+                   .AsSelf()
+                   .As(t => GetNonStandardBridgeInterfaces(t))
+                   .SingleInstance();
+
+            builder.RegisterType<DeviceProxy>()
+                   .AsSelf();
+
+            builder.RegisterType<SubscriptionChangeHandler>()
+                   .AsImplementedInterfaces()
+                   .SingleInstance();
+
+            builder.Register(c => new SystemComponentIdProvider(c.ResolveNamed<IClientCredentials>("EdgeHubCredentials")))
+                    .As<ISystemComponentIdProvider>()
+                    .SingleInstance();
+
+            builder.Register(
+                        c =>
+                        {
+                            var loggerFactory = c.Resolve<ILoggerFactory>();
+                            ILogger logger = loggerFactory.CreateLogger(typeof(MqttBridgeComponentDiscovery));
+
+                            var discovery = new MqttBridgeComponentDiscovery(logger);
+                            discovery.Discover(c);
+
+                            return discovery;
+                        })
+                .As<IComponentDiscovery>()
+                .SingleInstance();
+
+            builder.RegisterType<MqttBrokerConnector>()
+                   .AsImplementedInterfaces()
+                   .SingleInstance();
+
+            builder.Register(
+                        c =>
+                        {
+                            var connector = c.Resolve<IMqttBrokerConnector>();
+
+                            var port = this.config.GetValue("port", defaultPort);
+                            var baseUrl = this.config.GetValue("url", defaultUrl);
+
+                            var config = new MqttBrokerProtocolHeadConfig(port, baseUrl);
+
+                            return new MqttBrokerProtocolHead(config, connector);
+                        })
+                    .As<MqttBrokerProtocolHead>()
+                    .SingleInstance();
+
+            base.Load(builder);
+        }
+
+        static IEnumerable<Type> GetNonStandardBridgeInterfaces(Type type) => type.GetInterfaces().Where(t => !MqttBridgeComponentDiscovery.CandidateInterfaces.Contains(t));
```

Comment:
THis seems a bit hacky. What types implement multiple interfaces that need to be discovered?

Reply:
There are a couple of them. As I implemented handlers as separate classes, on of the 'not-so-good' side effect is that from the point of the DeviceProxy class, every operation is in a different class. Because of that, there are 4 classes injected to it (different handlers). These classes are registered by interfaces like "ITwinHandler or "ICloud2DeviceMessageHandler". In the same time, they must not be registered as "IMessageConsumer/Producer" - that is the filtering happening in the line above.